File: opcua_plugin/pymodules/pymqtt/plugin_main.py
# --coding:utf-8--
# Copyright (c) 2017 Intel Corporation
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#    http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.

# -----------------------------
# Copyright by Intel
# -----------------------------
from client import BasePluginClient
from entity import BasePluginEntity
from config import BasePluginConfig

import os
import time
import json
import ssl
import hashlib
import paho.mqtt.client as mqtt
from ret_codes import ReturnCodes
from logservice.logservice import LogService

logger = LogService.getLogger(__name__)


class MqttPluginEntity(BasePluginEntity):
    def __init__(self, file_name):
        super(MqttPluginEntity, self).__init__(file_name)

    def get_logger():
        return super(MqttPluginEntity, self).get_logger()


class MqttPluginConfig(BasePluginConfig):
    def __init__(self, config_fp):
        super(MqttPluginConfig, self).__init__(config_fp)


class MqttPluginClient(BasePluginClient):
    def __init__(self, entity, config):
        super(MqttPluginClient, self).__init__(entity, config)
        self.mqtt_devices = self.entity.get_custom_nodes()
        self.mqtt_clients = {}

    def _start(self):
        for device in self.mqtt_devices:
            mqttbroker = self.entity.get_property(device, 'MQTT Broker')
            if mqttbroker is None:
                continue
            broker_value = mqttbroker.value
            self.mqtt_clients[device.name] = {
                'name': device.name,
                'connected': False,
                'host': broker_value['host'] if 'host' in broker_value else '127.0.0.1',
                'port': broker_value['port'] if 'port' in broker_value else '1883',
                'user': broker_value['user'] if 'user' in broker_value else None,
                'password': broker_value['password'] if 'password' in broker_value else None,
                'cafile': broker_value['cafile'] if 'cafile' in broker_value else None,
                'cert': broker_value['cert'] if 'cert' in broker_value else None,
                'key': broker_value['key'] if 'key' in broker_value else None,
                'node_data': device,
                'client': None,
                'blob': None,
                'blob_hash': None
            }

        for c in self.mqtt_clients.values():
            self.mqtt_connect(c)

    def _stop(self):
        for c in self.mqtt_clients.values():
            self.mqtt_disconnect(c)

    def start(self):
        self._start()
        super(MqttPluginClient, self).start()

    def stop(self):
        self._stop()
        super(MqttPluginClient, self).stop()

    def getstate(self, name=None):
        return {
            'code': ReturnCodes.Good,
            'data': json.dumps(self.get_client_state(name))
        }

    def send(self, name, topic, msg):
        errorcode = 0
        try:
            c = self.getcli(name)
            if c is not None:
                if c['connected']:
                    c['client'].publish(topic, msg)

                    errorcode = 0
                    result = 'succeeded to send data'
                else:
                    errorcode = 1
                    result = 'the host name was not connected'
            else:
                errorcode = 1
                result = 'the host name was not initiated'
        except BaseException:
            errorcode = 1
            result = 'failed to send data'

        return {
            'code': errorcode,
            'data': result
        }

    def getcli(self, host_str):
        for client in self.mqtt_clients.values():
            if host_str == client['name']:
                return client
        return None

    def is_client_connected(self, name):
        if name:
            c = self.mqtt_clients.get(name, None)
            if c:
                return c['connected']
        return False

    def get_client_state(self, name):
        res = []
        client_list = []
        if name:
            c = self.mqtt_clients.get(name, None)
            if c:
                client_list.append(c)
        else:
            client_list = self.mqtt_clients.values()

        for client in client_list:
            res.append(json.dumps({
                'name': client['name'],
                'connected': client['connected'],
                'host': client['host'],
                'port': client['port'],
            }))

        return res

    def restart(self, name):
        if name:
            c = self.mqtt_clients.get(name, None)
            if c:
                self.mqtt_connect(c)
                self.mqtt_disconnect(c)
        else:
            self._stop()
            self._start()

    def plugin_poll(self):
        self.pub_event('', 'online', '')

    def mqtt_connect(self, mqtt_c):
        logger.debug('mqtt plugin connecting the broker......')
        node_data = mqtt_c['node_data']
        client = mqtt.Client(userdata=node_data)
        client.on_connect = self.mqtt_on_connect
        client.on_disconnect = self.mqtt_on_disconnect
        client.on_message = self.mqtt_on_message

        if mqtt_c['user'] is not None:
            client.username_pw_set(
                username=mqtt_c['user'],
                password=mqtt_c['password'])

        if mqtt_c['cafile'] is not None:
            if os.path.exists(
                    mqtt_c['cafile']) and os.path.exists(
                    mqtt_c['cert']) and os.path.exists(
                    mqtt_c['key']):
                client.tls_set(
                    ca_certs=mqtt_c['cafile'],
                    certfile=mqtt_c['cert'],
                    keyfile=mqtt_c['key'],
                    cert_reqs=ssl.CERT_REQUIRED,
                    tls_version=ssl.PROTOCOL_TLSv1,
                    ciphers=None)
            else:
                logger.info(
                    'The key files are not extists by (%s)' %
                    mqtt_c['name'])
        try:
            client.connect(mqtt_c['host'], int(mqtt_c['port']), 60)
        except BaseException:
            logger.info(
                'mqtt plugin connected the broker (%s) fail' %
                mqtt_c['name'])
        client.loop_start()
        mqtt_c['client'] = client

    def mqtt_disconnect(self, mqtt_c):
        if mqtt_c['client']:
            device_node = mqtt_c['node_data']
            datatopic = self.entity.get_property(device_node, 'Raw Topic')
            if datatopic is not None:
                mqtt_c['client'].unsubscribe(datatopic.value)

            alarmtopic = self.entity.get_property(device_node, 'Alam Topic')
            if alarmtopic is not None:
                mqtt_c['client'].unsubscribe(alarmtopic.value)

            blobtopic = self.entity.get_property(device_node, 'Blob Topic')
            if blobtopic is not None:
                mqtt_c['client'].unsubscribe(blobtopic.value)

            mqtt_c['client'].loop_stop()
            mqtt_c['client'].disconnect()

    def mqtt_on_connect(self, client, userdata, flags, rc):
        device_node = userdata
        datatopic = self.entity.get_property(device_node, 'Raw Topic')
        if datatopic is not None:
            client.subscribe(datatopic.value, 0)

        alarmtopic = self.entity.get_property(device_node, 'Alam Topic')
        if alarmtopic is not None:
            client.subscribe(alarmtopic.value, 0)

        blobtopic = self.entity.get_property(device_node, 'Blob Topic')
        if blobtopic is not None:
            client.subscribe(blobtopic.value, 0)

        self.mqtt_clients[device_node.name]['connected'] = True
        logger.info(
            'mqtt plugin connected the broker (%s) success' %
            device_node.name)

    def mqtt_on_disconnect(self, client, userdata, rc):
        device_node = userdata
        self.mqtt_clients[device_node.name]['connected'] = False
        logger.info(
            'mqtt plugin disconnect the broker (%s)' %
            device_node.name)

    def mqtt_on_message(self, client, userdata, msg):
        device_node = userdata

        blobtopic = self.entity.get_property(device_node, 'Blob Topic')
        if blobtopic is not None and blobtopic.value == msg.topic:
            self.on_blob_message(userdata, msg.payload)
        else:
            logger.debug(msg.topic + ":" + str(msg.payload.decode()))
            logger.debug(msg.topic + ":" + msg.payload.decode())

            message_str = msg.payload.decode()
            node = self.entity.get_node_by_val(msg.topic, device_node)
            pair_variable = node.get_pair_friend()
            self.pub_data(device_node.name, pair_variable.name, message_str)

    def on_blob_message(self, userdata, msg):
        device_node = userdata
        fout = self.mqtt_clients[device_node.name]['blob']
        if len(msg) == 0xFF:  # is header or end
            msg_in = msg.decode()
            msg_in = msg_in.split(":")
            if msg_in[0] == "END":
                in_hash_final = self.mqtt_clients[device_node.name]['blob_hash'].hexdigest(
                )
                if in_hash_final == msg_in[1]:
                    logger.debug("File Copied OK - Valid Hash")
                else:
                    logger.error("Bad File Received - Invalid Hash")
                fout.close()
                return
            elif msg_in[0] == "START":
                file_name = msg_in[1]
                file_path = os.path.split(os.path.realpath(self.entity.file_name))[
                    0] + '/' + file_name
                self.mqtt_clients[device_node.name]['blob'] = open(
                    file_path, "wb")
                self.mqtt_clients[device_node.name]['blob_hash'] = hashlib.md5()
                return
            else:
                pass
        self.mqtt_clients[device_node.name]['blob_hash'].update(msg)
        fout.write(msg)


def plugin_main(*args, **kwargs):
    plugin_entity = MqttPluginEntity(args[0])
    plugin_config = MqttPluginConfig(args[1])
    plugin_client = MqttPluginClient(plugin_entity, plugin_config)
    plugin_client.start()

File: opcua_plugin/javamodules/mqtt/src/main/java/com/intel/cfsdk/plugin/mqtt/PluginMain.java
package com.intel.cfsdk.plugin.mqtt;

import java.util.ArrayList;
import java.util.Date;
import java.util.HashMap;
import java.util.Iterator;
import java.util.List;
import java.util.Map;
import java.util.Properties;
import java.util.Random;
import java.io.BufferedReader;
import java.io.File;
import java.io.FileInputStream;
import java.io.IOException;
import java.io.InputStreamReader;

import org.eclipse.paho.client.mqttv3.MqttAsyncClient;
import org.eclipse.paho.client.mqttv3.MqttCallback;
import org.eclipse.paho.client.mqttv3.MqttClient;
import org.eclipse.paho.client.mqttv3.MqttConnectOptions;
import org.eclipse.paho.client.mqttv3.MqttDeliveryToken;
import org.eclipse.paho.client.mqttv3.MqttException;
import org.eclipse.paho.client.mqttv3.MqttMessage;
import org.eclipse.paho.client.mqttv3.MqttPersistenceException;
import org.eclipse.paho.client.mqttv3.MqttSecurityException;
import org.eclipse.paho.client.mqttv3.MqttTopic;
import org.eclipse.paho.client.mqttv3.persist.MemoryPersistence;

import net.sf.json.JSONArray;
import net.sf.json.JSONObject;

import com.intel.cfsdk.PluginClient;
import com.intel.cfsdk.Node;
import com.intel.cfsdk.plugin.mqtt.MqttPluginEntity;
import com.intel.cfsdk.plugin.mqtt.MqttPluginConfig;
import com.intel.cfsdk.PluginRpcImpl;

public class PluginMain extends PluginClient{
    private MqttPluginEntity devices;
    private PluginRpcImpl pluginRpcImpl;
    private JSONObject deviceClient;
    private JSONArray clients;
    private List<Map<String, Object>> deviceClients;

    public PluginMain(MqttPluginEntity entity, MqttPluginConfig config){
        super(entity, config);
        this.devices = entity;
        this.pluginRpcImpl = new PluginRpcImpl(this);
    }

    @Override
    public void registerRpc(){
        super.rpcNode.setCallback(this.pluginRpcImpl);
        try {
            super.rpcNode.connect();
        } catch (IOException e) {
            e.printStackTrace();
        }
    }
    
    public void pluginPoll(){
    	try {
			super.pubEvent("", "online", "");
		} catch (Exception e) {
			e.printStackTrace();
		}
    }
    
    public void mqtt_message(String topic,String msg){
    	for(int i = 0; i < this.deviceClients.size(); i++){
			Map<String, Object> map = this.deviceClients.get(i);
			Node blobTopic = ((Node) map.get("device")).getChild("Blob Topic");
			if(blobTopic != null && topic.equals(blobTopic.value)){
				System.out.println(blobTopic.value);
			}else{
				Node node = this.devices.getNodeByVal(topic, (Node)map.get("device"));
				if(node != null){
					Node pairNode = node.getPairFriend();
					try {
						super.pubData((String) map.get("name"), pairNode.name, msg);
					} catch (Exception e) {
						e.printStackTrace();
					}
				}
			}
    	}
    }
    
    public void mqtt_connect_sub(Map<String,Object> mapClient){
    	MqttClient mqttClient = (MqttClient) mapClient.get("client");
    	if(mqttClient.isConnected()){
    		Node cDevice = (Node) mapClient.get("device");
        	Node rawTopic = cDevice.getChild("Raw Topic");
        	if(rawTopic != null){
        		try {
    				mqttClient.subscribe(rawTopic.value, 1);
    			} catch (MqttException e) {
    				e.printStackTrace();
    			}
        	}
        	Node alamTopic = cDevice.getChild("Alam Topic");
        	if(alamTopic != null){
        		try {
    				mqttClient.subscribe(alamTopic.value, 1);
    			} catch (MqttException e) {
    				e.printStackTrace();
    			}
        	}
        	Node blobTopic = cDevice.getChild("Blob Topic");
        	if(blobTopic != null){
        		try {
    				mqttClient.subscribe(blobTopic.value, 1);
    			} catch (MqttException e) {
    				e.printStackTrace();
    			}
        	}
        	System.out.println("mqtt plugin connected the broker:" + mapClient.get("name"));
    	}
    }
    
    public void mqtt_connect(Map<String,Object> mapClient){
    	System.out.println("mqtt plugin connecting the broker......");
    	MqttClient mqttClient = null;
    	Properties props = null;
    	String clientId = MqttClient.generateClientId();
    	String ip = (String) mapClient.get("ip");
    	String path = this.getClass().getResource("").getPath();
    	path = path.split(":")[1];
    	path = path.substring(0, path.indexOf("!"));
    	path = path.substring(0, path.lastIndexOf("/")); 
    	//System.out.println(path);
    	String cafile = path + "/" + (String) mapClient.get("cafile");
    	String cert = path + "/" + (String) mapClient.get("cert");
    	String key = path + "/" + (String) mapClient.get("key");
    	
    	try {
    		mqttClient = new MqttClient(ip,clientId,new MemoryPersistence());
		} catch (MqttException e) {
			e.printStackTrace();
		}
    	MqttConnectOptions options = new MqttConnectOptions();
    	if((String) mapClient.get("cafile") != null){
    		if(new File(cafile).exists() && new File(cert).exists() && new File(key).exists()){
    			props.setProperty("cafile", cafile);
            	props.setProperty("cert", cert);
            	props.setProperty("key", key);
            	options.setSSLProperties(props);
    		}
    	}
    	options.setCleanSession(false);
    	options.setConnectionTimeout(10);
    	options.setKeepAliveInterval(20);
    	mqttClient.setCallback(new PushCallback(){
			public void messageArrived(String topic, MqttMessage message){
				String msg = new String(message.getPayload());
				System.out.println("message:" + msg);
				mqtt_message(topic,msg);
			}
		});
    	try {
    		mqttClient.connect(options);
    		Thread.sleep(1000);
		} catch (Exception e) {
			System.out.println("mqtt plugin connected the broker " + mapClient.get("name") + ":" + mqttClient.isConnected());
		}
    	mapClient.put("client", mqttClient);
    	mqtt_connect_sub(mapClient);
    }
    
    public void mqtt_disconnect(Map<String,Object> mapClient){
    	if(mapClient.get("client") != null){
    		Node cDevice = (Node) mapClient.get("device");
        	Node rawTopic = cDevice.getChild("Raw Topic");
        	if(rawTopic != null){
        		try {
        			((MqttClient) mapClient.get("client")).unsubscribe(rawTopic.value);
    			} catch (MqttException e) {
    				e.printStackTrace();
    			}
        	}
        	Node alamTopic = cDevice.getChild("Alam Topic");
        	if(alamTopic != null){
        		try {
        			((MqttClient) mapClient.get("client")).unsubscribe(alamTopic.value);
    			} catch (MqttException e) {
    				e.printStackTrace();
    			}
        	}
        	Node blobTopic = cDevice.getChild("Blob Topic");
        	if(blobTopic != null){
        		try {
        			((MqttClient) mapClient.get("client")).unsubscribe(blobTopic.value);
    			} catch (MqttException e) {
    				e.printStackTrace();
    			}
        	}
    	}
    	try {
			((MqttClient) mapClient.get("client")).disconnect();
		} catch (MqttException e) {
			e.printStackTrace();
		}
    	System.out.println("mqtt plugin disconnected the broker:" + mapClient.get("name"));
    }
    
    public Map<String,Object> getCli(String name){
    	if(name != null){
    		for(int i = 0; i < this.deviceClients.size(); i++){
    			Map<String, Object> map = this.deviceClients.get(i);
    			if(name.equals((String)map.get("name"))){
    				return map;
    			}
        	}
    	}
    	return null;
    }
    
    public JSONObject send(String name,String topic,String msg){
    	int errorcode = 0;
    	String result = null;
    	Map<String,Object> map = this.getCli(name);
    	if(map != null){
    		if(((MqttClient)map.get("client")).isConnected()){
    			try {
					((MqttClient)map.get("client")).publish(topic, msg.getBytes(), 1, false);
					errorcode = 0;
					result = "successed to send data";
				} catch (Exception e) {
					errorcode = 1;
					result = "failed to send data";
				}
    		}else{
    			errorcode = 1;
				result = "failed to send data";
    		}
    	}else{
    		errorcode = 1;
			result = "failed to send data";
    	}
    	Map<String, Object> res = new HashMap<String, Object>();
    	res.put("code", errorcode);
    	res.put("data", result);
    	return JSONObject.fromObject(res);
    }
    
    public void restart(String name){
    	Map<String,Object> map = this.getCli(name);
    	if(map != null){
    		this.mqtt_connect(map);
			this.mqtt_disconnect(map);
    	}else{
    		this.cli_stop();
    		this.cli_start();
    	}
    }
    
    public List<JSONObject> get_client_status(String name){
    	List<Map<String, Object>> list = new ArrayList<Map<String, Object>>();
    	Map<String, Object> listMap = new HashMap<String, Object>();
    	List<JSONObject> json = new ArrayList<JSONObject>();
    	Map<String,Object> map = this.getCli(name);
    	if(map != null){
    		list.add(map);
    	}else{
    		list = this.deviceClients;
    	}
    	for(int i = 0; i < list.size(); i++){
    		String cname = (String) list.get(i).get("name");
    		Boolean connected = ((MqttClient) list.get(i).get("client")).isConnected();
    		String ip = (String) list.get(i).get("ip");
    		listMap.put("name", cname);
    		listMap.put("connected", connected);
    		listMap.put("ip", ip);
    		json.add(JSONObject.fromObject(listMap));
    	}
    	return json;
    }
    
    public void cli_stop(){
    	for(int i = 0; i < this.deviceClients.size(); i++){
			Map<String, Object> map = this.deviceClients.get(i);
			this.mqtt_disconnect(map);
    	}
    }
    
    public void cli_start(){
    	List<Node> mqtt_devices = this.devices.getCustomNodes();
    	String ip, host, port, cafile, cert, key;
    	this.deviceClients = new ArrayList<Map<String, Object>>();
    	for(int i = 0; i < mqtt_devices.size(); i++){
    		Map<String, Object> mapClient = new HashMap<String, Object>();
    		MqttClient mqttClient = null;
    		Node device = mqtt_devices.get(i);
    		Node broker = device.getChild("MQTT Broker");
    		if(broker == null){
    			continue;
    		}
    		JSONObject broker_value = JSONObject.fromObject(broker.value);
    		if(broker_value.containsKey("host")){
    			host = broker_value.getString("host");
    		}else{
    			host = "127.0.0.1";
    		}
    		if(broker_value.containsKey("port")){
    			port = broker_value.getString("port");
    		}else{
    			port = "1883";
    		}
    		if(broker_value.containsKey("cafile")){
    			cafile = broker_value.getString("cafile");
    		}else{
    			cafile = null;
    		}
    		if(broker_value.containsKey("cert")){
    			cert = broker_value.getString("cert");
    		}else{
    			cert = null;
    		}
    		if(broker_value.containsKey("key")){
    			key = broker_value.getString("key");
    		}else{
    			key = null;
    		}
    		ip = "tcp://" + host + ":" + port;
    		mapClient.put("name", device.name);
    		mapClient.put("ip", ip);
    		mapClient.put("cafile", cafile);
    		mapClient.put("cert", cert);
    		mapClient.put("key", key);
    		mapClient.put("client", mqttClient);
    		mapClient.put("device", device);
    		this.deviceClients.add(mapClient);
    	}
    	for(int i=0;i<this.deviceClients.size();i++){
    		try {
				mqtt_connect(this.deviceClients.get(i));
			} catch (Exception e) {
				e.printStackTrace();
			}
    	}
    }
    
    public static void main( String[] args )
    {
        if (args.length < 1) {
            System.out.println("parameter error");
            return;
        }
        MqttPluginEntity pluginEntity = new MqttPluginEntity(args[0]);
        MqttPluginConfig pluginConfig = new MqttPluginConfig(args[1]);
        PluginMain pluginClient = new PluginMain(pluginEntity, pluginConfig);
        pluginClient.registerRpc();
        pluginClient.start(true);
        pluginClient.cli_start();
    }
}
